File: packages/less/src/less/tree/declaration.test.js
import { describe, it, expect, vi } from 'vitest';
import Declaration from './declaration';
import Value from './value';
import Keyword from './keyword';
import Anonymous from './anonymous';
import * as Constants from '../constants';

describe('Declaration', () => {
    describe('constructor', () => {
        it('creates a basic declaration', () => {
            const decl = new Declaration('color', 'red');
            expect(decl.name).toBe('color');
            expect(decl.value.value[0]).toBeInstanceOf(Anonymous);
            expect(decl.value.value[0].value).toBe('red');
            expect(decl.important).toBe('');
            expect(decl.merge).toBeUndefined();
            expect(decl.inline).toBe(false);
            expect(decl.variable).toBe(false);
        });

        it('handles variable declarations', () => {
            const decl = new Declaration('@color', 'red');
            expect(decl.variable).toBe(true);
        });

        it('handles important declarations', () => {
            const decl = new Declaration('color', 'red', '!important');
            expect(decl.important).toBe(' !important');
        });

        it('handles inline declarations', () => {
            const decl = new Declaration(
                'color',
                'red',
                null,
                false,
                0,
                {},
                true
            );
            expect(decl.inline).toBe(true);
        });

        it('accepts Node instances as values', () => {
            const valueNode = new Value([new Anonymous('blue')]);
            const decl = new Declaration('color', valueNode);
            expect(decl.value).toBe(valueNode);
        });

        it('handles null/undefined values', () => {
            const decl = new Declaration('color', null);
            expect(decl.value.value[0]).toBeNull();
        });

        it('handles array names for interpolation', () => {
            const keyword1 = new Keyword('border');
            const keyword2 = new Keyword('color');
            const decl = new Declaration([keyword1, keyword2], 'red');
            expect(Array.isArray(decl.name)).toBe(true);
            expect(decl.name[0]).toBe(keyword1);
            expect(decl.name[1]).toBe(keyword2);
        });

        it('handles merge flag', () => {
            const decl = new Declaration('color', 'red', null, true);
            expect(decl.merge).toBe(true);
        });

        it('preserves index and fileInfo', () => {
            const fileInfo = { filename: 'test.less' };
            const decl = new Declaration(
                'color',
                'red',
                null,
                false,
                123,
                fileInfo
            );
            expect(decl.getIndex()).toBe(123);
            expect(decl.fileInfo()).toBe(fileInfo);
        });
    });

    describe('genCSS', () => {
        it('generates basic CSS output', () => {
            const decl = new Declaration('color', 'red', null, false, 0, {});
            const output = { add: vi.fn() };
            decl.genCSS({ compress: false }, output);

            expect(output.add.mock.calls).toEqual([
                ['color: ', {}, 0],
                ['red', undefined, undefined, undefined],
                [';', {}, 0]
            ]);
        });

        it('generates compressed CSS output', () => {
            const decl = new Declaration('color', 'red', null, false, 0, {});
            const output = { add: vi.fn() };
            decl.genCSS({ compress: true }, output);

            expect(output.add.mock.calls).toEqual([
                ['color:', {}, 0],
                ['red', undefined, undefined, undefined],
                [';', {}, 0]
            ]);
        });

        it('generates CSS with important flag', () => {
            const decl = new Declaration(
                'color',
                'red',
                '!important',
                false,
                0,
                {}
            );
            const output = { add: vi.fn() };
            decl.genCSS({ compress: false }, output);

            expect(output.add.mock.calls).toEqual([
                ['color: ', {}, 0],
                ['red', undefined, undefined, undefined],
                [' !important;', {}, 0]
            ]);
        });

        it('omits semicolon for inline declarations', () => {
            const decl = new Declaration(
                'color',
                'red',
                null,
                false,
                0,
                {},
                true
            );
            const output = { add: vi.fn() };
            decl.genCSS({ compress: false }, output);

            expect(output.add.mock.calls).toEqual([
                ['color: ', {}, 0],
                ['red', undefined, undefined, undefined],
                ['', {}, 0]
            ]);
        });

        it('handles multiple values with commas', () => {
            const value = new Value([
                new Anonymous('1px'),
                new Anonymous('solid'),
                new Anonymous('black')
            ]);
            const decl = new Declaration('border', value, null, false, 0, {});
            const output = { add: vi.fn() };
            decl.genCSS({ compress: false }, output);

            expect(output.add.mock.calls).toEqual([
                ['border: ', {}, 0],
                ['1px', undefined, undefined, undefined],
                [', '],
                ['solid', undefined, undefined, undefined],
                [', '],
                ['black', undefined, undefined, undefined],
                [';', {}, 0]
            ]);
        });

        it('handles compressed output with multiple values', () => {
            const value = new Value([
                new Anonymous('1px'),
                new Anonymous('solid'),
                new Anonymous('black')
            ]);
            const decl = new Declaration('border', value, null, false, 0, {});
            const output = { add: vi.fn() };
            decl.genCSS({ compress: true }, output);

            expect(output.add.mock.calls).toEqual([
                ['border:', {}, 0],
                ['1px', undefined, undefined, undefined],
                [','],
                ['solid', undefined, undefined, undefined],
                [','],
                ['black', undefined, undefined, undefined],
                [';', {}, 0]
            ]);
        });

        it('handles last rule in compressed mode', () => {
            const decl = new Declaration('color', 'red', null, false, 0, {});
            const output = { add: vi.fn() };
            decl.genCSS({ compress: true, lastRule: true }, output);

            expect(output.add.mock.calls[2][0]).toBe(''); // No semicolon for last rule
        });

        it('handles errors in value generation', () => {
            const badValue = new Value([
                {
                    genCSS: () => {
                        throw new Error('Test error');
                    },
                    value: 'bad'
                }
            ]);
            const decl = new Declaration('color', badValue, null, false, 0, {
                filename: 'test.less'
            });
            const output = { add: vi.fn() };

            expect(() => decl.genCSS({ compress: false }, output)).toThrow(
                'Test error'
            );
        });
    });

    describe('eval', () => {
        const createContext = () => ({
            importantScope: [],
            math: Constants.Math.PARENS_DIVISION
        });

        it('evaluates simple declarations', () => {
            const anonymous = new Anonymous('red', 0, {
                filename: 'test.less'
            });
            expect(anonymous.value).toBe('red');

            const value = new Value([anonymous]);
            const decl = new Declaration('color', value, null, false, 0, {
                filename: 'test.less'
            });
            const context = createContext();
            const evaluated = decl.eval(context);

            expect(evaluated).toBeInstanceOf(Declaration);
            expect(evaluated.name).toBe('color');
            expect(evaluated.value).toBeInstanceOf(Anonymous);
            expect(evaluated.value.value).toBe('red');
        });

        it('evaluates interpolated names', () => {
            const keyword1 = new Keyword('border');
            const anonymous = new Anonymous('-');
            const keyword2 = new Keyword('color');
            const decl = new Declaration(
                [keyword1, anonymous, keyword2],
                'red'
            );
            const context = createContext();
            const evaluated = decl.eval(context);

            expect(evaluated.name).toBe('border-color');
        });

        it('handles detached ruleset errors', () => {
            const detachedValue = {
                type: 'DetachedRuleset',
                eval: () => ({ type: 'DetachedRuleset' })
            };
            const value = new Value([detachedValue]);
            const decl = new Declaration('prop', value, null, false, 0, {
                filename: 'test.less'
            });

            expect(() => decl.eval(createContext())).toThrow(
                'Rulesets cannot be evaluated on a property'
            );
        });

        it('handles font declarations with math context', () => {
            const decl = new Declaration('font', 'bold');
            const context = {
                ...createContext(),
                math: Constants.Math.ALWAYS
            };
            const evaluated = decl.eval(context);

            expect(evaluated.name).toBe('font');
            expect(context.math).toBe(Constants.Math.ALWAYS);
        });

        it('restores math context after font evaluation', () => {
            const decl = new Declaration('font', 'bold');
            const context = {
                ...createContext(),
                math: Constants.Math.ALWAYS
            };
            decl.eval(context);

            expect(context.math).toBe(Constants.Math.ALWAYS);
        });

        it('evaluates declarations with keyword names', () => {
            const keyword = new Keyword('width');
            const decl = new Declaration([keyword], 'auto');
            const context = createContext();
            const evaluated = decl.eval(context);

            expect(evaluated.name).toBe('width');
        });

        it('preserves important flags during evaluation', () => {
            const decl = new Declaration('color', 'red', '!important');
            const context = createContext();
            const evaluated = decl.eval(context);

            expect(evaluated.important).toBe(' !important');
        });

        it('handles errors during evaluation', () => {
            const badValue = new Value([
                {
                    eval: () => {
                        throw new Error('Eval error');
                    }
                }
            ]);
            const decl = new Declaration('color', badValue, null, false, 0, {
                filename: 'test.less'
            });
            const context = createContext();

            expect(() => decl.eval(context)).toThrow('Eval error');
        });
    });

    describe('makeImportant', () => {
        it('creates a new declaration with important flag', () => {
            const decl = new Declaration('color', 'red');
            const important = decl.makeImportant();

            expect(important).toBeInstanceOf(Declaration);
            expect(important.name).toBe('color');
            expect(important.value).toBe(decl.value);
            expect(important.important).toBe(' !important');
        });

        it('preserves other properties when making important', () => {
            const decl = new Declaration(
                'color',
                'red',
                null,
                true,
                1,
                { filename: 'test.less' },
                true
            );
            const important = decl.makeImportant();

            expect(important.merge).toBe(true);
            expect(important.getIndex()).toBe(1);
            expect(important.fileInfo()).toEqual({ filename: 'test.less' });
            expect(important.inline).toBe(true);
        });
    });
});

File: packages/less/src/less/go_parser/declaration_test.go
package go_parser

import (
	"testing"

	"github.com/toakleaf/less.go/packages/less/src/less"
)

func TestDeclaration(t *testing.T) {
	t.Run("constructor", func(t *testing.T) {
		t.Run("creates a basic declaration", func(t *testing.T) {
			decl, err := NewDeclaration("color", "red", nil, false, 0, nil, false, nil)
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if decl.name != "color" {
				t.Errorf("expected name to be 'color', got %v", decl.name)
			}
			if _, ok := decl.value.value[0].(*Anonymous); !ok {
				t.Error("expected value to be Anonymous")
			}
			if decl.value.value[0].(*Anonymous).Value != "red" {
				t.Errorf("expected value to be 'red', got %v", decl.value.value[0].(*Anonymous).Value)
			}
			if decl.important != "" {
				t.Errorf("expected important to be empty, got %v", decl.important)
			}
			if decl.merge {
				t.Error("expected merge to be false")
			}
			if decl.inline {
				t.Error("expected inline to be false")
			}
			if decl.variable {
				t.Error("expected variable to be false")
			}
		})

		t.Run("handles variable declarations", func(t *testing.T) {
			decl, err := NewDeclaration("@color", "red", nil, false, 0, nil, false, nil)
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if !decl.variable {
				t.Error("expected variable to be true")
			}
		})

		t.Run("handles important declarations", func(t *testing.T) {
			decl, err := NewDeclaration("color", "red", "!important", false, 0, nil, false, nil)
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if decl.important != " !important" {
				t.Errorf("expected important to be ' !important', got %v", decl.important)
			}
		})

		t.Run("handles inline declarations", func(t *testing.T) {
			decl, err := NewDeclaration("color", "red", nil, false, 0, nil, true, nil)
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if !decl.inline {
				t.Error("expected inline to be true")
			}
		})

		t.Run("accepts Value instances", func(t *testing.T) {
			anonymousValue := NewAnonymous("blue", 0, nil, false, false, nil)
			value, err := NewValue([]any{anonymousValue})
			if err != nil {
				t.Fatalf("unexpected error creating value: %v", err)
			}
			decl, err := NewDeclaration("color", value, nil, false, 0, nil, false, nil)
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if decl.value != value {
				t.Error("expected value to be the provided Value instance")
			}
		})

		t.Run("handles null/undefined values", func(t *testing.T) {
			decl, err := NewDeclaration("color", nil, nil, false, 0, nil, false, nil)
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if decl.value.value[0] != nil {
				t.Error("expected value to be nil")
			}
		})

		t.Run("handles array names for interpolation", func(t *testing.T) {
			keyword1 := NewKeyword("border")
			keyword2 := NewKeyword("color")
			decl, err := NewDeclaration([]any{keyword1, keyword2}, "red", nil, false, 0, nil, false, nil)
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			nameArr, ok := decl.name.([]any)
			if !ok {
				t.Fatal("expected name to be array")
			}
			if len(nameArr) != 2 {
				t.Fatalf("expected name array length 2, got %d", len(nameArr))
			}
			if nameArr[0] != keyword1 {
				t.Error("expected first name element to be keyword1")
			}
			if nameArr[1] != keyword2 {
				t.Error("expected second name element to be keyword2")
			}
		})
	})

	t.Run("genCSS", func(t *testing.T) {
		t.Run("generates basic CSS output", func(t *testing.T) {
			decl, _ := NewDeclaration("color", "red", nil, false, 0, nil, false, nil)
			var output []string
			cssOutput := &CSSOutput{
				Add: func(chunk any, fileInfo any, index any) {
					if chunk != nil {
						output = append(output, chunk.(string))
					}
				},
			}
			decl.GenCSS(map[string]any{"compress": false}, cssOutput)

			expected := []string{"color", ": ", "red", ";"}
			if len(output) != len(expected) {
				t.Fatalf("expected output length %d, got %d", len(expected), len(output))
			}
			for i, v := range expected {
				if output[i] != v {
					t.Errorf("at index %d expected %q, got %q", i, v, output[i])
				}
			}
		})

		t.Run("generates compressed CSS output", func(t *testing.T) {
			decl, _ := NewDeclaration("color", "red", nil, false, 0, nil, false, nil)
			var output []string
			cssOutput := &CSSOutput{
				Add: func(chunk any, fileInfo any, index any) {
					if chunk != nil {
						output = append(output, chunk.(string))
					}
				},
			}
			decl.GenCSS(map[string]any{"compress": true}, cssOutput)

			expected := []string{"color", ":", "red", ";"}
			if len(output) != len(expected) {
				t.Fatalf("expected output length %d, got %d", len(expected), len(output))
			}
			for i, v := range expected {
				if output[i] != v {
					t.Errorf("at index %d expected %q, got %q", i, v, output[i])
				}
			}
		})

		t.Run("generates CSS with important flag", func(t *testing.T) {
			decl, _ := NewDeclaration("color", "red", "!important", false, 0, nil, false, nil)
			var output []string
			cssOutput := &CSSOutput{
				Add: func(chunk any, fileInfo any, index any) {
					if chunk != nil {
						output = append(output, chunk.(string))
					}
				},
			}
			decl.GenCSS(map[string]any{"compress": false}, cssOutput)

			expected := []string{"color", ": ", "red", " !important", ";"}
			if len(output) != len(expected) {
				t.Fatalf("expected output length %d, got %d", len(expected), len(output))
			}
			for i, v := range expected {
				if output[i] != v {
					t.Errorf("at index %d expected %q, got %q", i, v, output[i])
				}
			}
		})

		t.Run("handles multiple values with commas", func(t *testing.T) {
			value, _ := NewValue([]any{
				NewAnonymous("1px", 0, nil, false, false, nil),
				NewAnonymous("solid", 0, nil, false, false, nil),
				NewAnonymous("black", 0, nil, false, false, nil),
			})
			decl, _ := NewDeclaration("border", value, nil, false, 0, nil, false, nil)
			var output []string
			cssOutput := &CSSOutput{
				Add: func(chunk any, fileInfo any, index any) {
					if chunk != nil {
						output = append(output, chunk.(string))
					}
				},
			}
			decl.GenCSS(map[string]any{"compress": false}, cssOutput)

			expected := []string{"border", ": ", "1px", ", ", "solid", ", ", "black", ";"}
			if len(output) != len(expected) {
				t.Fatalf("expected output length %d, got %d", len(expected), len(output))
			}
			for i, v := range expected {
				if output[i] != v {
					t.Errorf("at index %d expected %q, got %q", i, v, output[i])
				}
			}
		})

		t.Run("handles compressed output with multiple values", func(t *testing.T) {
			value, _ := NewValue([]any{
				NewAnonymous("1px", 0, nil, false, false, nil),
				NewAnonymous("solid", 0, nil, false, false, nil),
				NewAnonymous("black", 0, nil, false, false, nil),
			})
			decl, _ := NewDeclaration("border", value, nil, false, 0, nil, false, nil)
			var output []string
			cssOutput := &CSSOutput{
				Add: func(chunk any, fileInfo any, index any) {
					if chunk != nil {
						output = append(output, chunk.(string))
					}
				},
			}
			decl.GenCSS(map[string]any{"compress": true}, cssOutput)

			expected := []string{"border", ":", "1px", ",", "solid", ",", "black", ";"}
			if len(output) != len(expected) {
				t.Fatalf("expected output length %d, got %d", len(expected), len(output))
			}
			for i, v := range expected {
				if output[i] != v {
					t.Errorf("at index %d expected %q, got %q", i, v, output[i])
				}
			}
		})

		t.Run("omits semicolon for inline declarations", func(t *testing.T) {
			decl, _ := NewDeclaration("color", "red", nil, false, 0, nil, true, nil)
			var output []string
			cssOutput := &CSSOutput{
				Add: func(chunk any, fileInfo any, index any) {
					if chunk != nil {
						output = append(output, chunk.(string))
					}
				},
			}
			decl.GenCSS(map[string]any{"compress": false}, cssOutput)

			expected := []string{"color", ": ", "red", ""}
			if len(output) != len(expected) {
				t.Fatalf("expected output length %d, got %d", len(expected), len(output))
			}
			for i, v := range expected {
				if output[i] != v {
					t.Errorf("at index %d expected %q, got %q", i, v, output[i])
				}
			}
		})

		t.Run("handles last rule in compressed mode", func(t *testing.T) {
			decl, _ := NewDeclaration("color", "red", nil, false, 0, nil, false, nil)
			var output []string
			cssOutput := &CSSOutput{
				Add: func(chunk any, fileInfo any, index any) {
					if chunk != nil {
						output = append(output, chunk.(string))
					}
				},
			}
			decl.GenCSS(map[string]any{"compress": true, "lastRule": true}, cssOutput)

			expected := []string{"color", ":", "red", ""}
			if len(output) != len(expected) {
				t.Fatalf("expected output length %d, got %d", len(expected), len(output))
			}
			for i, v := range expected {
				if output[i] != v {
					t.Errorf("at index %d expected %q, got %q", i, v, output[i])
				}
			}
		})

		t.Run("handles errors in value generation", func(t *testing.T) {
			badValue := &Value{
				value: []any{
					&ErrorGeneratingValue{},
				},
			}
			decl, _ := NewDeclaration("color", badValue, nil, false, 0, map[string]any{"filename": "test.less"}, false, nil)
			cssOutput := &CSSOutput{
				Add: func(chunk any, fileInfo any, index any) {},
			}

			defer func() {
				if r := recover(); r == nil {
					t.Error("expected panic from error generating value")
				}
			}()
			decl.GenCSS(map[string]any{"compress": false}, cssOutput)
		})
	})

	t.Run("eval", func(t *testing.T) {
		createContext := func() map[string]any {
			return map[string]any{
				"importantScope": []any{},
				"math":          less.Math.ParensDivision,
			}
		}

		t.Run("evaluates simple declarations", func(t *testing.T) {
			anonymous := NewAnonymous("red", 0, map[string]any{"filename": "test.less"}, false, false, nil)
			value, _ := NewValue([]any{anonymous})
			decl, _ := NewDeclaration("color", value, nil, false, 0, map[string]any{"filename": "test.less"}, false, nil)
			
			evaluated, err := decl.Eval(createContext())
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			
			evalDecl, ok := evaluated.(*Declaration)
			if !ok {
				t.Fatal("expected evaluated value to be Declaration")
			}
			if evalDecl.name != "color" {
				t.Errorf("expected name to be 'color', got %v", evalDecl.name)
			}
		})

		t.Run("evaluates interpolated names", func(t *testing.T) {
			keyword1 := NewKeyword("border")
			anonymous := NewAnonymous("-", 0, nil, false, false, nil)
			keyword2 := NewKeyword("color")
			decl, _ := NewDeclaration([]any{keyword1, anonymous, keyword2}, "red", nil, false, 0, nil, false, nil)
			
			evaluated, err := decl.Eval(createContext())
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			
			evalDecl, ok := evaluated.(*Declaration)
			if !ok {
				t.Fatal("expected evaluated value to be Declaration")
			}
			if evalDecl.name != "border-color" {
				t.Errorf("expected name to be 'border-color', got %v", evalDecl.name)
			}
		})

		t.Run("handles font declarations with math context", func(t *testing.T) {
			decl, _ := NewDeclaration("font", "bold", nil, false, 0, nil, false, nil)
			context := createContext()
			context["math"] = less.Math.Always
			
			evaluated, err := decl.Eval(context)
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			
			evalDecl, ok := evaluated.(*Declaration)
			if !ok {
				t.Fatal("expected evaluated value to be Declaration")
			}
			if evalDecl.name != "font" {
				t.Errorf("expected name to be 'font', got %v", evalDecl.name)
			}
			if context["math"] != less.Math.Always {
				t.Error("expected math context to be restored")
			}
		})

		t.Run("handles detached ruleset errors", func(t *testing.T) {
			detachedValue := &Value{
				value: []any{
					map[string]any{"type": "DetachedRuleset"},
				},
			}
			decl, _ := NewDeclaration("prop", detachedValue, nil, false, 0, map[string]any{"filename": "test.less"}, false, false)
			
			_, err := decl.Eval(createContext())
			if err == nil {
				t.Error("expected error for detached ruleset")
			}
			if err.Error() != "Rulesets cannot be evaluated on a property" {
				t.Errorf("unexpected error message: %v", err)
			}
		})

		t.Run("handles errors during evaluation", func(t *testing.T) {
			badValue := &Value{
				value: []any{
					&ErrorEvaluatingValue{},
				},
			}
			decl, _ := NewDeclaration("color", badValue, nil, false, 0, map[string]any{"filename": "test.less"}, false, nil)
			
			_, err := decl.Eval(createContext())
			if err == nil {
				t.Error("expected error during evaluation")
			}
			if err.Error() != "eval error" {
				t.Errorf("unexpected error message: %v", err)
			}
		})
	})

	t.Run("makeImportant", func(t *testing.T) {
		t.Run("creates a new declaration with important flag", func(t *testing.T) {
			decl, _ := NewDeclaration("color", "red", nil, false, 0, nil, false, nil)
			important := decl.MakeImportant()

			if important.name != "color" {
				t.Errorf("expected name to be 'color', got %v", important.name)
			}
			if important.value != decl.value {
				t.Error("expected value to be preserved")
			}
			if important.important != " !important" {
				t.Errorf("expected important to be ' !important', got %v", important.important)
			}
		})

		t.Run("preserves other properties", func(t *testing.T) {
			fileInfo := map[string]any{"filename": "test.less"}
			decl, _ := NewDeclaration("color", "red", nil, true, 1, fileInfo, true, nil)
			important := decl.MakeImportant()

			if !important.merge {
				t.Error("expected merge to be preserved")
			}
			if important.GetIndex() != 1 {
				t.Errorf("expected index to be preserved, got %v", important.GetIndex())
			}
			if important.FileInfo()["filename"] != "test.less" {
				t.Error("expected fileInfo to be preserved")
			}
			if !important.inline {
				t.Error("expected inline to be preserved")
			}
		})
	})
}

// Helper types for error testing
type ErrorGeneratingValue struct{}

func (e *ErrorGeneratingValue) GenCSS(context any, output *CSSOutput) {
	panic("Test error")
}

type ErrorEvaluatingValue struct{}

func (e *ErrorEvaluatingValue) Eval(context any) (any, error) {
	return nil, &Error{Message: "eval error"}
} 